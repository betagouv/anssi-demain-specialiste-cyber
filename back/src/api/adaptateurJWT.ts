import jwt, { JwtPayload } from 'jsonwebtoken';

export interface AdaptateurJWT {
  genereToken(donnees: Record<string, unknown>): string;
  decode(token: string): JwtPayload;
}

export const adaptateurJWT: AdaptateurJWT = {
  genereToken: (donnees: Record<string, unknown>) =>
    jwt.sign(donnees, process.env.SECRET_JWT || '', { expiresIn: '1h' }),
  decode: (token: string) =>
    jwt.verify(token, process.env.SECRET_JWT || '') as JwtPayload,
};
