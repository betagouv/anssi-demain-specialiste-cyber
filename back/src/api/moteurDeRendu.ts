import express from 'express';
import fs from 'fs';
import { join } from 'path';

export interface MoteurDeRendu {
  rends: (reponse: express.Response, vue: string, options?: object) => void;
}

type FournisseurDeChemins = () => { dsc: string; style: string };

const lisLesCheminsDesFichiersGeneresParLeFront: FournisseurDeChemins = () => {
  const cheminVersLeFichierManifeste = join(
    process.cwd(),
    '../front/dist/.vite/manifest.json',
  );
  const manifeste = JSON.parse(
    fs.readFileSync(cheminVersLeFichierManifeste, 'utf-8'),
  );
  return {
    dsc: manifeste['src/index.ts'].file,
    style: manifeste['style.css'].file,
  };
};

export const moteurDeRenduExpress = (
  fournisseurDeChemins: FournisseurDeChemins = lisLesCheminsDesFichiersGeneresParLeFront,
): MoteurDeRendu => {
  return {
    rends(reponse, vue, options) {
      const optionsAvecManifesteEtNonce = {
        ...options,
        ...fournisseurDeChemins(),
        nonce: reponse.locals.nonce,
      };
      reponse.render(vue, optionsAvecManifesteEtNonce);
    },
  };
};
