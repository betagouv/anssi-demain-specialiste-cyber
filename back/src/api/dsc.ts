import { ConfigurationServeurLab, creeServeurLab } from '@lab-anssi/lib';
import cookieParser from 'cookie-parser';
import cookieSession from 'cookie-session';
import express from 'express';
import { AdaptateurHachage } from '../infra/adaptateurHachage';
import { RecupereCheminVersFichiersStatiques } from '../infra/recupereCheminVersFichiersStatiques';
import { EntrepotRessourcesCyber } from '../metier/entrepotRessourcesCyber';
import { EntrepotUtilisateur } from '../metier/entrepotUtilisateur';
import { AdaptateurJWT } from './adaptateurJWT';
import { AdaptateurOIDC } from './oidc/adaptateurOIDC';
import { ressourceApresAuthentificationOIDC } from './oidc/ressourceApresAuthentificationOIDC';
import { ressourceConnexionOIDC } from './oidc/ressourceConnexionOIDC';
import { ressourceRessourceCyber } from './ressourceRessourcesCyber';

export interface ConfigurationServeur {
  serveurLab: ConfigurationServeurLab;
  entrepotRessourcesCyber: EntrepotRessourcesCyber;
  adaptateurOIDC: AdaptateurOIDC;
  adaptateurJWT: AdaptateurJWT;
  entrepotUtilisateur: EntrepotUtilisateur;
  adaptateurHachage: AdaptateurHachage;
  recupereCheminVersFichiersStatiques: RecupereCheminVersFichiersStatiques;
}

export const creeServeur = (configurationServeur: ConfigurationServeur) => {
  const app = creeServeurLab(configurationServeur.serveurLab);

  app.use(
    cookieSession({
      name: 'session',
      sameSite: true,
      secret: process.env.SECRET_COOKIE,
      secure: process.env.NODE_ENV === 'production',
      signed: process.env.NODE_ENV === 'production',
    })
  );

  app.use(cookieParser());

  app.use(
    '/',
    express.static(configurationServeur.recupereCheminVersFichiersStatiques())
  );

  app.use('/oidc/connexion', ressourceConnexionOIDC(configurationServeur));
  app.use(
    '/oidc/apres-authentification',
    ressourceApresAuthentificationOIDC(configurationServeur)
  );

  app.use(
    '/api/ressources-cyber',
    ressourceRessourceCyber(configurationServeur)
  );

  return app;
};
