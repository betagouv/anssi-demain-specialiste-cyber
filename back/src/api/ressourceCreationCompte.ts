import { Router } from 'express';
import { ConfigurationServeur } from './dsc';

export const ressourceCreationCompte = ({
  adaptateurJWT,
  adaptateurRechercheEntreprise,
  moteurDeRendu,
}: ConfigurationServeur) => {
  const routeur = Router();

  routeur.get('/', async (requete, reponse) => {
    const token = requete.query.token;
    if (!token) return reponse.sendStatus(400);

    const informationsProfessionnelles = adaptateurJWT.decode(token as string);

    const organisation =
      await adaptateurRechercheEntreprise.rechercheOrganisationParSiret(
        informationsProfessionnelles.siret
      );

    moteurDeRendu.rends(reponse, 'creation-compte', {
      informationsProfessionnelles: {
        nom: informationsProfessionnelles.nom,
        prenom: informationsProfessionnelles.prenom,
        email: informationsProfessionnelles.email,
        organisation: {
          departement: organisation?.departement,
          nom: organisation?.nom,
          siret: informationsProfessionnelles.siret,
        },
      },
    });
  });

  return routeur;
};
