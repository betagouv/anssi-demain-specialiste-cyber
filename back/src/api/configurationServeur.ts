import { ConfigurationServeurLab } from '@lab-anssi/lib';
import { BusEvenements } from '../bus/busEvenements';
import { AdaptateurEnvironnement } from '../infra/adaptateurEnvironnement';
import { AdaptateurHachage } from '../infra/adaptateurHachage';
import { AdaptateurJournal } from '../infra/adaptateurJournal';
import { AdaptateurRechercheEntreprise } from '../infra/adaptateurRechercheEntreprise';
import { RecupereCheminVersFichiersStatiques } from '../infra/recupereCheminVersFichiersStatiques';
import { EntrepotJeux } from '../metier/entrepotJeux';
import { EntrepotRessourcesCyber } from '../metier/entrepotRessourcesCyber';
import { EntrepotUtilisateur } from '../metier/entrepotUtilisateur';
import { AdaptateurJWT } from './adaptateurJWT';
import { Middleware } from './middleware';
import { MoteurDeRendu } from './moteurDeRendu';
import { AdaptateurOIDC } from './oidc/adaptateurOIDC';
import { AdaptateurTeleversement } from '../infra/adaptateurTeleversement';

export interface ConfigurationServeur {
  adaptateurEnvironnement: AdaptateurEnvironnement;
  serveurLab: ConfigurationServeurLab;
  entrepotRessourcesCyber: EntrepotRessourcesCyber;
  adaptateurOIDC: AdaptateurOIDC;
  adaptateurJWT: AdaptateurJWT;
  adaptateurRechercheEntreprise: AdaptateurRechercheEntreprise;
  entrepotUtilisateur: EntrepotUtilisateur;
  adaptateurHachage: AdaptateurHachage;
  recupereCheminsVersFichiersStatiques: RecupereCheminVersFichiersStatiques;
  middleware: Middleware;
  moteurDeRendu: MoteurDeRendu;
  busEvenements: BusEvenements;
  entrepotJeux: EntrepotJeux;
  adaptateurJournal: AdaptateurJournal;
  adaptateurTeleversement: AdaptateurTeleversement;
}

export type ConfigurationServeurSansMiddleware = Omit<
  ConfigurationServeur,
  'middleware'
>;
