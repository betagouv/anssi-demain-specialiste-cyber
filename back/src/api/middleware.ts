import { HttpStatusCode } from 'axios';
import { NextFunction, Request, Response } from 'express';
import { randomBytes } from 'node:crypto';
import z from 'zod';
import { ConfigurationServeurSansMiddleware } from './configurationServeur';
import { AdaptateurHachage } from '../infra/adaptateurHachage';
import { EntrepotUtilisateur } from '../metier/entrepotUtilisateur';
import { Utilisateur } from '../metier/utilisateur';

type FonctionMiddleware<TBody> = (
  requete: Request<unknown, unknown, TBody, unknown, never>,
  reponse: Response,
  suite: NextFunction,
) => Promise<void>;

export type Middleware = {
  verifieModeMaintenance: FonctionMiddleware<unknown>;
  valideLaCoherenceDuCorps: <
    TZod extends z.ZodType,
    TBody extends z.infer<TZod>,
  >(
    object: TZod,
  ) => FonctionMiddleware<TBody>;
  verifieJWTNavigation: FonctionMiddleware<unknown>;
  ajouteUtilisateurARequete(
    entrepotUtilisateur: EntrepotUtilisateur,
    adaptateurHachage: AdaptateurHachage,
  ): FonctionMiddleware<unknown>;
};

export type RequeteNonTypee = Request<
  unknown,
  unknown,
  unknown,
  unknown,
  never
>;

export const fabriqueMiddleware = ({
  adaptateurEnvironnement,
  adaptateurJWT,
}: ConfigurationServeurSansMiddleware): Middleware => {
  const verifieModeMaintenance = async (
    _requete: RequeteNonTypee,
    reponse: Response,
    suite: NextFunction,
  ) => {
    if (adaptateurEnvironnement.maintenance().actif()) {
      const nonce = randomBytes(24).toString('base64');
      reponse
        .status(HttpStatusCode.ServiceUnavailable)
        .set('Content-Type', 'text/html')
        .render('maintenance', { nonce });
    } else {
      suite();
    }
  };

  const valideLaCoherenceDuCorps =
    <TZod extends z.ZodType, TBody extends z.infer<TZod>>(objet: TZod) =>
    async (
      requete: Request<unknown, unknown, TBody, unknown, never>,
      reponse: Response,
      suite: NextFunction,
    ) => {
      const resultat = objet.safeParse(requete.body);
      if (!resultat.success) {
        reponse.status(400).json({ erreur: resultat.error.issues[0].message });
      } else {
        suite();
      }
    };

  const verifieJWTNavigation = async (
    requete: RequeteNonTypee,
    reponse: Response,
    suite: NextFunction,
  ) => {
    if (!requete.session?.token) {
      return reponse.redirect('/connexion');
    }
    try {
      adaptateurJWT.decode(requete.session.token);
      suite();
    } catch {
      reponse.redirect('/connexion');
    }
  };

  const ajouteUtilisateurARequete =
    (
      entrepotUtilisateur: EntrepotUtilisateur,
      adaptateurHachage: AdaptateurHachage,
    ) =>
    async (
      requete: RequeteNonTypee & { utilisateur?: Utilisateur | undefined },
      reponse: Response,
      suite: NextFunction,
    ) => {
      try {
        const { email } = adaptateurJWT.decode(requete.session?.token);
        requete.utilisateur = email
          ? await entrepotUtilisateur.parEmailHache(
              adaptateurHachage.hache(email),
            )
          : undefined;
        suite();
      } catch {
        reponse.sendStatus(500);
      }
    };

  return {
    verifieModeMaintenance,
    valideLaCoherenceDuCorps,
    verifieJWTNavigation,
    ajouteUtilisateurARequete,
  };
};
