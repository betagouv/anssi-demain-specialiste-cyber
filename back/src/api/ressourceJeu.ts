import { Router } from 'express';
import z from 'zod';
import { Jeu } from '../metier/jeu';
import { ConfigurationServeur } from './configurationServeur';

type ReponseJeu = Omit<
  Jeu,
  'enseignant' | 'incrementeReaction' | 'decrementeReaction'
> & {
  enseignant: string;
};

const schemaModificationJeu = z.strictObject({
  estCache: z.boolean().optional(),
});

export const ressourceJeu = ({
  entrepotJeux,
  middleware,
}: ConfigurationServeur) => {
  const routeur = Router();

  routeur.get('/:id', async (requete, reponse) => {
    const jeu = await entrepotJeux.parId(requete.params.id);
    if (!jeu) {
      return reponse.sendStatus(404);
    }
    reponse.send(enReponseJeu(jeu));
  });

  routeur.patch(
    '/:id',
    middleware.valideLaCoherenceDuCorps(schemaModificationJeu),
    async (requete, reponse) => {
      const jeu = await entrepotJeux.parId(
        (requete.params as Record<string, string>).id,
      );
      if (!jeu) {
        return reponse.sendStatus(404);
      }
      if (requete.body.estCache !== undefined) {
        jeu.estCache = requete.body.estCache;
      }
      await entrepotJeux.metsAjour(jeu);
      reponse.status(200).send(enReponseJeu(jeu));
    },
  );

  return routeur;
};

const enReponseJeu = (jeu: Jeu): ReponseJeu => ({
  ...jeu,
  enseignant: jeu.enseignant?.prenom ?? '',
});
