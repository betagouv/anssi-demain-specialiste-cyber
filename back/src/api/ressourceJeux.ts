import { Router } from 'express';
import z from 'zod';
import { JeuCree } from '../bus/evenements/jeu/jeuCree';
import { Jeu } from '../metier/jeu';
import { ConfigurationServeur } from './configurationServeur';
import { sequences } from '../metier/referentiels/sequence';
import { classes } from '../metier/referentiels/classes';
import { disciplines } from '../metier/referentiels/disciplines';

function chaineNonVide(message: string) {
  return z.string(message).trim().min(1, message);
}

export const ressourceJeux = ({
  entrepotJeux,
  entrepotUtilisateur,
  adaptateurHachage,
  middleware,
  busEvenements,
}: ConfigurationServeur) => {
  const routeur = Router();

  const schema = z.strictObject({
    nom: chaineNonVide('Le nom est obligatoire'),
    nomEtablissement: chaineNonVide(
      'Le nom de l‘établissement est obligatoire',
    ),
    discipline: z.enum(disciplines, {
      error: 'La discipline est invalide',
    }),
    classe: z.enum(classes, {
      error: 'La classe est invalide',
    }),
    sequence: z.enum(sequences, {
      error: 'La séquence est invalide',
    }),
    eleves: z
      .array(chaineNonVide('Les prénoms fournis sont invalides'))
      .nonempty('Au moins un élève est requis'),
  });

  routeur.post(
    '/',
    middleware.valideLaCoherenceDuCorps(schema),
    middleware.ajouteUtilisateurARequete(
      entrepotUtilisateur,
      adaptateurHachage,
    ),
    async (requete, reponse) => {
      try {
        const utilisateurConnecte = requete.utilisateur;

        const { nom, sequence, nomEtablissement, classe, discipline, eleves } =
          requete.body;
        await entrepotJeux.ajoute(
          new Jeu({
            nom,
            enseignant: utilisateurConnecte,
            sequence,
            nomEtablissement,
            classe,
            discipline,
            eleves,
          }),
        );
        await busEvenements.publie(
          new JeuCree(
            utilisateurConnecte.email,
            nom,
            sequence,
            nomEtablissement,
            classe,
            discipline,
          ),
        );
        reponse.sendStatus(201);
      } catch {
        reponse.sendStatus(401);
      }
    },
  );

  routeur.get(
    '/',
    middleware.ajouteUtilisateurARequete(
      entrepotUtilisateur,
      adaptateurHachage,
    ),
    async (requete, reponse) => {
      const jeux = await entrepotJeux.lesJeuxDe(requete.utilisateur);
      reponse.send(jeux.map((jeu) => ({ id: jeu.id, nom: jeu.nom })));
    },
  );

  return routeur;
};
