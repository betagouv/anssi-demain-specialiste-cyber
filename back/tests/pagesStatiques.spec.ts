import { Express } from 'express';
import request from 'supertest';
import { describe, expect, it } from 'vitest';
import { creeServeur } from '../src/api/dsc';
import { configurationDeTestDuServeur } from './api/fauxObjets';

describe('Le serveur de pages statiques', () => {
  const serveur = creeServeur(configurationDeTestDuServeur());

  const appelleUnePageStatique = (serveur: Express) =>
    request(serveur).get('/');

  it('répond 200', async () => {
    const reponse = await appelleUnePageStatique(serveur);

    expect(reponse.status).toEqual(200);
  });

  it('répond avec un contenu HTML', async () => {
    const reponse = await appelleUnePageStatique(serveur);

    expect(reponse.headers['content-type']).toEqual('text/html; charset=UTF-8');
  });

  it('répond avec le contenu HTML', async () => {
    const reponse = await appelleUnePageStatique(serveur);

    expect(reponse.text).toMatchSnapshot();
  });
});
