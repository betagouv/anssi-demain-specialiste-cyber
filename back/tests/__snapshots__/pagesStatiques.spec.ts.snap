// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Le serveur de pages statiques > r√©pond avec le contenu HTML 1`] = `
"<!doctype html>
<html>
  <body>
    Catalogue
  </body>
</html>
"
`;
